python/src/trezorlib/cli/settings.py

```diff
@@ -60,15 +66,15 @@ def image_to_tt(filename: str) -> bytes:
         except Exception as e:
             raise click.ClickException("TOIF file is corrupted") from e
 
-    elif Image is None:
+    elif not PIL_AVAILABLE:
         raise click.ClickException(
             "Image library is missing. Please install via 'pip install Pillow'."
         )
 
     else:
         try:
             image = Image.open(filename)
-            toif_image = toif.from_image(image)
+            toif_image = toif.from_image(image)  # type: ignore [Argument of type "Image" cannot be assigned to parameter "image"]
```

Comment:
this type:ignore is weird. isn't it solved by the `Image.Image` thing in toif.py?

related: does pyright report unused type:ignores?

Reply:
Correct, it is no longer necessary after those `Image.Image` changes.

I could not find this option for `pyright`. There is one issue asking for it, but was closed without implementing - the main developer of it did not like it - https://github.com/microsoft/pyright/issues/450#issuecomment-568198054 - but he has good points

Reply:
Type:ignore deleted in b70acaf

Reply:
i disagree with that comment strongly .) but whatever.

we should make a point of periodically running with "type:ignore" disabled and checking if the resulting errors match the type:ignored comments.

In fact, it might be possible to write our own type:ignore feature.
Set disableTypeIgnore and then run the following wrapper:
1. run pyright and tally the errors
2. extract source lines on which errors occur
3. if there's a type:ignore, match the error to what is in brackets
4. if the error matches, ignore it and exclude from output

not in this PR, but if you're interested, you can try

btw how many type:ignores are there in the codebase right now?

Reply:
In `trezorctl` there are currently just 9 `type: ignore`s and all of them are valid now (checked with `"enableTypeIgnoreComments": false`). 

Funny thing is that the version of `pyright` was increased from `1.1.187` to `1.1.190` and it caught 6 new errors:
```
/home/jmusil/trezor-firmware/python/src/trezorlib/_ed25519.py
  /home/jmusil/trezor-firmware/python/src/trezorlib/_ed25519.py:88:9 - error: Operator "*" not supported for types "NoReturn" and "NoReturn"
    Operator "*" not supported for types "NoReturn" and "NoReturn" (reportGeneralTypeIssues)
  /home/jmusil/trezor-firmware/python/src/trezorlib/_ed25519.py:89:14 - error: Operator "*" not supported for types "NoReturn" and "NoReturn"
    Operator "*" not supported for types "NoReturn" and "NoReturn" (reportGeneralTypeIssues)
  /home/jmusil/trezor-firmware/python/src/trezorlib/_ed25519.py:91:8 - error: Operator "%" not supported for types "Unknown | NoReturn" and "Literal[2]"
    Operator "%" not supported for types "NoReturn" and "Literal[2]" (reportGeneralTypeIssues)
  /home/jmusil/trezor-firmware/python/src/trezorlib/_ed25519.py:94:12 - error: Expression of type "Unknown | Any | NoReturn" cannot be assigned to return type "int"
    Type "Unknown | Any | NoReturn" cannot be assigned to type "int"
      "NoReturn" is incompatible with "int" (reportGeneralTypeIssues)
/home/jmusil/trezor-firmware/python/src/trezorlib/protobuf.py
  /home/jmusil/trezor-firmware/python/src/trezorlib/protobuf.py:40:43 - error: Function with declared type of "int" must return value
    Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
  /home/jmusil/trezor-firmware/python/src/trezorlib/protobuf.py:48:36 - error: Function with declared type of "int" must return value
```


The purpose of that wrapper would be just to have that `"reportUnusedTypeIgnores"`? Or would we use it to do line-specific ignores specifying the exact error type (like `reportMissingParameterType`) - which is currently not available as far a I know?

Reply:
In `core`, we do have 111 `type:ignore`s for mypy, and we do have `warn_unused_ignores` in `setup.cfg`.

For `pyright` it could be different, we should get rid of all of them and create a specific ignore message in `[]`.

What comes to my mind is that one line can sometimes have multiple issues at once - but we can have some separator like `;;` in the ignore brackets - `# type: ignore [QObject" is possibly unbound;;"SIGNAL" is possibly unbound]` - real example from `python/src/trezorlib/qt/pinmatrix.py:161`

Reply:
Theoretically we do not need to use `# type: ignore` either, we could have `# pyright: ignore [...]`, so we are not messing up with `mypy` or other type-checkers.

With that own tool we could easily ignore specific errors in specific files, and even in specific functions, or having something like `# pyright: off` and `# pyright: on` marks to disable whole code block.

Implementation should not be so hard, there is a `CLI` option `pyright --outputjson` which has all the info in easy-to-analyze structure

Reply:
all good ideas.

wrt `pinmatrix`, i'm kind of tempted to exclude the whole file

it would be great to get rid of 3rd party library problems (hid, webusb) in strict mode. we could do that relatively easily via own stubs, it seems.